### PR TITLE
ATO-579: create identity progress lambda

### PR DESCRIPTION
**Pull request by isaac-GDS**
Identity Progress lambda and exposed on frontend api

### CODE REVIEW

File: ci/terraform/oidc/api-gateway-frontend.tf [hunk 1]
```diff
@@ -86,6 +86,7 @@ resource "aws_api_gateway_deployment" "frontend_deployment" {
       local.deploy_check_email_fraud_block_count == 1 ? module.check_email_fraud_block[0].integration_trigger_value : null,
       local.deploy_check_email_fraud_block_count == 1 ? module.check_email_fraud_block[0].method_trigger_value : null,
       local.account_modifiers_encryption_policy_arn,
```
Comment (hannah-leong) [inline ci/terraform/oidc/api-gateway-frontend.tf:88]: Maybe add commas on the end to match the pattern before

File: ci/terraform/oidc/api-gateway-frontend.tf [hunk 2]
```diff
@@ -86,6 +86,7 @@ resource "aws_api_gateway_deployment" "frontend_deployment" {
...
+      var.orch_identity_progress_enabled
```
Comment (hannah-leong) [inline ci/terraform/oidc/api-gateway-frontend.tf:89]:
```suggestion
      var.orch_identity_progress_enabled,
```

File: ci/terraform/oidc/api-gateway-frontend.tf [hunk 3]
```diff
@@ -110,6 +111,7 @@ resource "aws_api_gateway_deployment" "frontend_deployment" {
     module.doc-app-authorize,
     module.orch_auth_code,
     module.check_reauth_user,
+    aws_api_gateway_integration.orch_identity_progress_integration
```
Comment (hannah-leong) [inline ci/terraform/oidc/api-gateway-frontend.tf:114]:
```suggestion
    aws_api_gateway_integration.orch_identity_progress_integration,
```

File: ci/terraform/oidc/api-gateway.tf [hunk 4]
```diff
@@ -1283,4 +1282,4 @@ resource "aws_api_gateway_integration" "orch_userinfo_integration" {
   type                    = "AWS_PROXY"
   integration_http_method = "POST"
   uri                     = "arn:aws:apigateway:eu-west-2:lambda:path/2015-03-31/functions/arn:aws:lambda:eu-west-2:${var.orch_account_id}:function:${var.orch_userinfo_name}:latest/invocations"
-}
+}
```
Comment (hannah-leong) [inline ci/terraform/oidc/api-gateway.tf:1285]: Leave the new line in?

File: ci/terraform/oidc/api-gateway-frontend.tf [hunk 5]
```diff
@@ -358,3 +360,38 @@ resource "aws_wafv2_web_acl_logging_configuration" "waf_logging_config_frontend_
...
+  rest_api_id = aws_api_gateway_rest_api.di_authentication_frontend_api.id
+  resource_id = aws_api_gateway_resource.orch_identity_progress_resource[0].id
+  http_method = aws_api_gateway_method.orch_identity_progress_method[0].http_method
+  depends_on = [
+    aws_api_gateway_resource.orch_identity_progress_resource
+  ]
+  type                    = "AWS_PROXY"
+  integration_http_method = "POST"
+  uri                     = "arn:aws:apigateway:eu-west-2:lambda:path/2015-03-31/functions/arn:aws:lambda:eu-west-2:${var.orch_account_id}:function:${var.orch_identity_progress_name}:latest/invocations"
+}
```
Comment (hannah-leong) [inline ci/terraform/oidc/api-gateway-frontend.tf:397]: Leave the new line in?

File: template.yaml [hunk 6]
```diff
@@ -3217,6 +3222,208 @@ Resources:
...
+              - dynamodb:Query
+              - dynamodb:Scan
+            Resource: !Sub
+              - 'arn:aws:dynamodb:eu-west-2:${AccountId}:table/${AuthEnvironment}-identity-credentials'
+              - AccountId: !FindInMap [ EnvironmentConfiguration, !Ref Environment, authAccountId ]
+                AuthEnvironment: !FindInMap [ EnvironmentConfiguration, !Ref Environment, authEnvironment ]
+          - Sid: AllowDecrpytUserInfoTable
+            Effect: "Allow"
+            Action: "kms:Decrypt*"
+            Resource:
```
Comment (hannah-leong) [inline template.yaml:3312]: For consistency
```suggestion
          - Sid: AllowDecrpytUserInfoTable
            Effect: Allow
            Action: kms:Decrypt*
            Resource:
```

File: template.yaml [hunk 7]
```diff
@@ -3217,6 +3222,208 @@ Resources:
...
+              - kms:DescribeKey
+            Resource: !FindInMap [ EnvironmentConfiguration, !Ref Environment, txmaEventEncryptionKeyArn ]
+          - Sid: AllowSQSTxMA
+            Effect: Allow
+            Action:
+              - sqs:SendMessage
+            Resource:
+              - !Sub arn:aws:sqs:eu-west-2:${AWS::AccountId}:orchestration-${Environment}-txma-AuditEventQueue
+          - Sid: AllowUserInfoTableReadAccess
+            Effect: "Allow"
```
Comment (hannah-leong) [inline template.yaml:3281]: For consistency:
```suggestion
            Effect: Allow
```

File: template.yaml [hunk 8]
```diff
@@ -3217,6 +3222,208 @@ Resources:
...
+              - "ssm:GetParameters"
+            Resource:
+              - !Sub "arn:aws:ssm:eu-west-2:${AWS::AccountId}:parameter/${Environment}-session-redis-master-host"
+              - !Sub "arn:aws:ssm:eu-west-2:${AWS::AccountId}:parameter/${Environment}-session-redis-tls"
+              - !Sub "arn:aws:ssm:eu-west-2:${AWS::AccountId}:parameter/${Environment}-session-redis-port"
+              - !Sub "arn:aws:ssm:eu-west-2:${AWS::AccountId}:parameter/${Environment}-session-redis-password"
+          - Sid: AllowDecryptOfParameters
+            Effect: "Allow"
+            Action: "kms:Decrypt"
+            Resource: !GetAtt MainKmsKey.Arn
```
Comment (hannah-leong) [inline template.yaml:3327]: More for consistency

**Review by hannah-leong — COMMENTED** [review]
Looks good, just some stylistic change requests :)

**Comment by isaac-GDS** [comment]
Rebased to main and updated changes but might be worth checking anything i might have missed.

Things updates: 
- Moved away from function name env
- Added checkgov for ATO-657

File: template.yaml [hunk 9]
```diff
@@ -3339,6 +3344,209 @@ Resources:
...
+              - !Sub "arn:aws:ssm:eu-west-2:${AWS::AccountId}:parameter/${Environment}-session-redis-master-host"
+              - !Sub "arn:aws:ssm:eu-west-2:${AWS::AccountId}:parameter/${Environment}-session-redis-tls"
+              - !Sub "arn:aws:ssm:eu-west-2:${AWS::AccountId}:parameter/${Environment}-session-redis-port"
+              - !Sub "arn:aws:ssm:eu-west-2:${AWS::AccountId}:parameter/${Environment}-session-redis-password"
+          - Sid: AllowDecryptOfParameters
+            Effect: Allow
+            Action: kms:Decrypt
+            Resource: !GetAtt MainKmsKey.Arn
+      Tags:
+        CheckovRulesToSkip: CKV_AWS_116.CKV_AWS_117.CKV_AWS_173
```
Comment (hannah-leong) [inline template.yaml:3452]: Think 115 should also be here?

File: template.yaml [hunk 10]
```diff
@@ -3339,6 +3344,209 @@ Resources:
...
+          - Sid: AllowGetParameters
+            Effect: Allow
+            Action:
+              - "ssm:GetParameter"
+              - "ssm:GetParameters"
+            Resource:
+              - !Sub "arn:aws:ssm:eu-west-2:${AWS::AccountId}:parameter/${Environment}-session-redis-master-host"
+              - !Sub "arn:aws:ssm:eu-west-2:${AWS::AccountId}:parameter/${Environment}-session-redis-tls"
+              - !Sub "arn:aws:ssm:eu-west-2:${AWS::AccountId}:parameter/${Environment}-session-redis-port"
+              - !Sub "arn:aws:ssm:eu-west-2:${AWS::AccountId}:parameter/${Environment}-session-redis-password"
```
Comment (hannah-leong) [inline template.yaml:3446]: Just being picky, but maybe remove quotation marks from these for consistency?

**Review by hannah-leong — COMMENTED** [review]
Might be worth checking the code scanning comments and dismiss them if they don't apply (haven't looked at them all but looks like some can be dismissed)

File: template.yaml [hunk 11]
```diff
@@ -3339,6 +3344,209 @@ Resources:
...
+              - dynamodb:BatchGetItem
+              - dynamodb:DescribeTable
+              - dynamodb:Get*
+              - dynamodb:Query
+              - dynamodb:Scan
+            Resource: !Sub
+              - 'arn:aws:dynamodb:eu-west-2:${AccountId}:table/${AuthEnvironment}-identity-credentials'
+              - AccountId: !FindInMap [ EnvironmentConfiguration, !Ref Environment, authAccountId ]
+                AuthEnvironment: !FindInMap [ EnvironmentConfiguration, !Ref Environment, authEnvironment ]
+          - Sid: AllowDecrpytUserInfoTable
```
Comment (hannah-leong) [inline template.yaml:3432]:
```suggestion
          - Sid: AllowDecryptUserInfoTable
```

File: template.yaml (hunk 9 above)
Comment (isaac-GDS) [reply template.yaml:3452]: Good spot!

File: template.yaml (hunk 10 above)
Comment (isaac-GDS) [reply template.yaml:3446]: Good to be picky :)

File: template.yaml (hunk 11 above)
Comment (isaac-GDS) [reply template.yaml:3432]: Looks like copied over the wrong spelling - good find

**Comment by isaac-GDS** [comment]
Rebased - switched userInfoEncryptionKeyArn to authenticationCallbackUserinfoKeyArn

something to look for is if any current policies could be referenced

File: template.yaml [hunk 12]
```diff
@@ -2374,6 +2379,149 @@ Resources:
...
+        - 'arn:aws:execute-api:eu-west-2:${AccountId}:${FrontendApiId}/*/GET/identity-progress'
+        - AccountId: !FindInMap [ EnvironmentConfiguration, !Ref Environment, authAccountId ]
+          FrontendApiId: !FindInMap [ EnvironmentConfiguration, !Ref Environment, authFrontendApiId ]
+
+  IdentityProgressFunctionLogGroup:
+    Type: AWS::Logs::LogGroup
+    Properties:
+      LogGroupName: !Sub /aws/lambda/${Environment}-identity-progress-lambda
+      KmsKeyId: !GetAtt MainKmsKey.Arn
+      RetentionInDays: 14
```
[the remark this answers is not in the document]
Comment (isaac-GDS) [reply template.yaml:2445]: Confirmed that we should be defaulting to 30days. Thanks @kalpaitch for bringing it up. Will update this lambda here and the other lambdas in a separate PR.

**Comment by ethanmills** [comment]
@isaac-GDS Shall we close this for now and revisit when the orch frontend is closer?

**Comment by isaac-GDS** [comment]
> @isaac-GDS Shall we close this for now and revisit when the orch frontend is closer?

Its ready to be merged - but happy to close to avoid unused code in prod.

**Pull request closed without merging**